prometheus/prometheus.go

```diff
@@ -150,19 +179,27 @@ func (p *PrometheusSink) AddSample(parts []string, val float32) {
 
 func (p *PrometheusSink) AddSampleWithLabels(parts []string, val float32, labels []metrics.Label) {
 	key, hash := p.flattenKey(parts, labels)
-	g, ok := p.summaries.Load(hash)
-	if !ok {
-		g = prometheus.NewSummary(prometheus.SummaryOpts{
+	ps, ok := p.summaries.Load(hash)
+
+	if ok {
+		localSummary := *ps.(*PrometheusSummary)
+		localSummary.Observe(float64(val))
+		localSummary.updatedAt = time.Now()
+		p.summaries.Store(hash, &localSummary)
```

Comment:
I'm a little concerned that this drops concurrent updates-- you've solved a "physical" race condition but moved it up a layer to become a "logical" race condition, if that analogy makes any sense?

I don't have a good alternative recommendation, though, perhaps serializing the updates through a channel?  Doing a compare-and-swap?

Reply:
Hmm, you're right.  I'd persuaded myself that it was ok in the counter/gauge cases, but I'd forgotten about summaries.  I'll have to think more about this.

Reply:
Now I'm not so sure you're right.  We're making a copy of a PrometheusSummary, but its Summary field is an interface.  So we're really just copying a pointer to a struct, and thus if a parallel goroutine were to also call AddSampleWithLabels, both goroutines that reach these lines would be calling Observe on the same Summary instance.

We could still lose updates in the case where there was no entry in the p.summaries map, so each goroutine would create its own Summary and only one would get it to store it.  That's unfortunate, though not new to this PR (mea culpa, it was brought in with my PR that introduced sync.Map and the panic this PR solves).  That should happen much more rarely though.  And since we're setting MaxAge=10s we have a much bigger problem with lost observations IMO.  (I'll follow up with a PR to address that after this much more urgent panic is addressed.)

Reply:
OK, got it.  (It was not clear to me that prometheus.Summary was an interface type.)

I see that the implementation type prometheus.summary has its own mutex.